### PR TITLE
WebSocket integration

### PR DESCRIPTION
**Pull request by amaury901130**
#### Description:

* Web Socket integration.

### CODE REVIEW

File: RSDemoProject/Chat/ViewModels/ChatViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+//
+//  Created by Amaury Ricardo on 6/23/20.
+//  Copyright © 2020 TopTier labs. All rights reserved.
+//
+
+import Foundation
+
+enum ChatState {
+  case chatMessagesLoaded
+  case messageSended
```
Comment (glm4) [inline RSDemoProject/Chat/ViewModels/ChatViewModel.swift:13]: should be `.messageSent`

File: RSDemoProject/Managers/WebSocketManager.swift [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+//  RSDemoProject
+//
+//  Created by Amaury Ricardo on 6/23/20.
+//  Copyright © 2020 TopTier labs. All rights reserved.
+//
+
+import Foundation
+import Starscream
+
+class WSS: WebSocketDelegate {
```
Comment (glm4) [inline RSDemoProject/Managers/WebSocketManager.swift:12]: avoid abbreviations

File: RSDemoProject/Managers/WebSocketManager.swift [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  }
+  
+  private var baseHeaders: [String: String] {
+    return [
+      HTTPHeader.accept.rawValue: "application/json",
+      HTTPHeader.contentType.rawValue: "application/json"
+    ]
+  }
+  
+  var headers: [String: String] {
```
Comment (glm4) [inline RSDemoProject/Managers/WebSocketManager.swift:27]: can't you use headers from `APIClient`?

File: RSDemoProject/Managers/WebSocketManager.swift [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      return baseHeaders + [
+        HTTPHeader.uid.rawValue: session.uid ?? "",
+        HTTPHeader.client.rawValue: session.client ?? "",
+        HTTPHeader.token.rawValue: session.accessToken ?? ""
+      ]
+    }
+    return baseHeaders
+  }
+  
+  var ws: WebSocket!
```
Comment (glm4) [inline RSDemoProject/Managers/WebSocketManager.swift:38]: this variable should be private, also avoid abbreviations.

File: RSDemoProject/Managers/WebSocketManager.swift [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  ) -> String? {
+    let message: [String : String] = [
+      "action": "send_message",
+      "content": content,
+      "match_conversation_id": "\(conversationId)"
+    ]
+    
+    let jsonData = try? JSONEncoder().encode(message)
+    
+    guard let data = jsonData else {
```
Comment (glm4) [inline RSDemoProject/Managers/WebSocketManager.swift:81]: join this conditions`, guard let data = try? ...`

File: RSDemoProject/Managers/WebSocketManager.swift [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    case .disconnected(let reason, _):
+      delegate?.onDisconnected(reason)
+    case .text(let string):
+      delegate?.onText(string)
+    default: break
+    }
+  }
+}
+
+protocol WSSDelegate {
```
Comment (glm4) [inline RSDemoProject/Managers/WebSocketManager.swift:101]: avoid abbreviations.

File: RSDemoProject/Managers/WebSocketManager.swift [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  func onConnected()
+  func onDisconnected(_ reason: String?)
+  func onConnectionError(_ reason: String)
+  func onText(_ text: String)
```
Comment (glm4) [inline RSDemoProject/Managers/WebSocketManager.swift:105]: use more _swifty_ and informative names like: socketDidConnect, onMessageReceived, etc

File: RSDemoProject/Managers/WebSocketManager.swift [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    
+    ws.write(string: message, completion: completion)
+  }
+  
+  private func getFormattedMessage(
+    content: String,
+    conversationId: Int
+  ) -> String? {
+    let message: [String : String] = [
```
Comment (glm4) [inline RSDemoProject/Managers/WebSocketManager.swift:73]: associate the colon to the identifier type `[String: String]`

File: RSDemoProject/Chat/ViewModels/ChatViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+class ChatViewModel {
+  
+  init(_ conversation: Conversation) {
+    self.conversation = conversation
+  }
+  
+  var conversation: Conversation!
+  var messages: [Message] = []
+  var page = 1
+  var webSocketManager = WSS.shared
```
Comment (germanStabile) [inline RSDemoProject/Chat/ViewModels/ChatViewModel.swift:31]: this should be a let

File: RSDemoProject/Managers/WebSocketManager.swift (hunk 2 above)
Comment (germanStabile) [inline RSDemoProject/Managers/WebSocketManager.swift:12]: maybe `WebSocketManager`?

File: RSDemoProject/Managers/WebSocketManager.swift [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  var headers: [String: String] {
+    if let session = SessionManager.currentSession {
+      return baseHeaders + [
+        HTTPHeader.uid.rawValue: session.uid ?? "",
+        HTTPHeader.client.rawValue: session.client ?? "",
+        HTTPHeader.token.rawValue: session.accessToken ?? ""
+      ]
+    }
+    return baseHeaders
+  }
```
Comment (germanStabile) [inline RSDemoProject/Managers/WebSocketManager.swift:36]: Aren't these the exact same headers you are using in the APIClient?

File: RSDemoProject/Managers/WebSocketManager.swift (hunk 5 above)
Comment (germanStabile) [inline RSDemoProject/Managers/WebSocketManager.swift:81]: you can join this with line 79:
```
guard let data = try? JSONEncoder().encode(message) ...
```

File: RSDemoProject/Managers/WebSocketManager.swift (hunk 3 above)
Comment (amaury901130) [reply RSDemoProject/Managers/WebSocketManager.swift:27]: The problem that it is implemented in an extension of the protocol TargetType and that added an unuseful code to the WebSocket class.

Comment (glm4) [reply RSDemoProject/Managers/WebSocketManager.swift:27]: You can use `APIClient.baseHeaders` for the first one and `APIClient.getHeaders()` with some modifications for the second one. We must avoid code repetition always.

**Pull request merged**